In [8]:
%load_ext sql


The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [9]:
%config SqlMagic.autocommit=False # for engines that do not support autommit

In [10]:
%sql select * from campaign_info.csv

Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])
Can't load plugin: sqlalchemy.dialects:presto
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


In [33]:
import pandas as pd
from pandasql import sqldf
campaign = pd.read_csv(r"campaign_info.csv")

marketing = pd.read_csv(r"marketing_performance.csv")

website_rev = pd.read_csv(r"website_revenue.csv")

In [34]:
campaign.head(5)

id       name   status   last_updated_date
0  64441550  Campaign1   paused  2023-08-08 0:00:00
1  55304737  Campaign2  enabled  2023-07-25 0:00:00
2  89363211  Campaign3  enabled  2023-08-06 0:00:00
3  86363015  Campaign4   paused  2023-08-14 0:00:00
4  99058240  Campaign5  enabled  2023-07-28 0:00:00

In [35]:
marketing.head(5)

date  campaign_id               geo    cost  impressions  \
0  2023-07-29 0:00:00     64441550  United States-TX  107.86         2486   
1  2023-08-14 0:00:00     64441550  United States-OH  230.99         2465   
2  2023-08-12 0:00:00     64441550  United States-GA  214.94         2461   
3  2023-08-04 0:00:00     55304737  United States-NY  123.74         1069   
4  2023-08-19 0:00:00     86363015  United States-GA  122.74         1248   

   clicks  conversions  
0    1737          294  
1    1747          275  
2    1306          352  
3    1356          317  
4    1290          388

In [80]:
website_rev.head(5)

date  campaign_id state  revenue
0  2023-07-24 0:00:00     64441550    GA     6370
1  2023-08-02 0:00:00     89363211    CA     6676
2  2023-08-03 0:00:00     86363015    CA     6239
3  2023-07-24 0:00:00     55304737    TX     6607
4  2023-08-17 0:00:00     86363015    NY     7014

**1. Write a query to get the sum of impressions by day.**

In [79]:

sqldf('''
SELECT date, SUM(impressions) as total_daily_impressions
FROM marketing
GROUP BY date
ORDER BY date
''')

date  total_daily_impressions
0   2023-07-24 0:00:00                     1423
1   2023-07-26 0:00:00                     1547
2   2023-07-27 0:00:00                     2295
3   2023-07-28 0:00:00                     8142
4   2023-07-29 0:00:00                     3511
5   2023-07-30 0:00:00                     3001
6   2023-08-02 0:00:00                     1532
7   2023-08-03 0:00:00                     3189
8   2023-08-04 0:00:00                     3254
9   2023-08-05 0:00:00                     1088
10  2023-08-10 0:00:00                     2128
11  2023-08-11 0:00:00                     2278
12  2023-08-12 0:00:00                     6588
13  2023-08-13 0:00:00                     2170
14  2023-08-14 0:00:00                     5773
15  2023-08-17 0:00:00                     3292
16  2023-08-19 0:00:00                     1248

**2. Write a query to get the top three revenue-generating states in order of best to worst. How much revenue did the third best state generate?**


In [54]:
sqldf('''
SELECT state, revenue
FROM website_rev
GROUP BY state
ORDER BY revenue desc
LIMIT 3
''')

state  revenue
0    NY     7014
1    CA     6676
2    TX     6607

The third best state Texas (TX) generated 6607 in revenue.

**3. Write a query that shows total cost, impressions, clicks, and revenue of each campaign. Make sure to include the campaign name in the output.**

In [64]:
sqldf('''
SELECT name, SUM(cost) as total_cost, impressions, clicks, revenue
FROM campaign
JOIN marketing
  ON campaign.id = marketing.campaign_id
JOIN website_rev
  ON campaign.id = website_rev.campaign_id
 GROUP BY name

''')

name  total_cost  impressions  clicks  revenue
0  Campaign1     4170.51         1532    1806     5571
1  Campaign2     4075.50         1069    1356     5824
2  Campaign3    15809.04         2295    1139     5249
3  Campaign4     3970.14         2469    1617     6239
4  Campaign5     4077.15         1025    1724     5130

**4. Write a query to get the number of conversions of Campaign5 by state. Which state generated the most conversions for this campaign?**

In [78]:
#marketing.coversions
sqldf('''
SELECT name, geo, SUM(conversions) as Total_Conversions
FROM campaign
JOIN marketing
  ON campaign.id = marketing.campaign_id
WHERE name = 'Campaign5'
 GROUP BY geo
 ORDER BY Total_Conversions DESC
''')

name               geo  Total_Conversions
0  Campaign5  United States-GA                672
1  Campaign5  United States-OH                442

The state of Georgia (GA) generated the most conversions at 672.

**5. In your opinion, which campaign was the most efficient, and why?**


There are many ways to define efficient, so it depends. Let's take most efficient to mean most conversions by lowest cost per unit, then Campaign 4 is the most efficient. 
See the results below for the most efficient campaign to the least efficient campaign

In [81]:
sqldf('''
SELECT name, SUM(cost) as total_cost, conversions, SUM(cost)/conversions as cost_per_conversion
FROM campaign
JOIN marketing
  ON campaign.id = marketing.campaign_id
JOIN website_rev
  ON campaign.id = website_rev.campaign_id
 GROUP BY name
 ORDER BY cost_per_conversion ASC

''')

name  total_cost  conversions  cost_per_conversion
0  Campaign4     3970.14          499             7.956192
1  Campaign1     4170.51          470             8.873426
2  Campaign5     4077.15          346            11.783671
3  Campaign2     4075.50          317            12.856467
4  Campaign3    15809.04          287            55.083763

**BONUS QUESTION! Write a query that showcases the best day of the week (e.g., Sunday, Monday, Tuesday, etc.) to run ads.**


In [111]:
sqldf('''
SELECT AVG(clicks), case cast(STRFTIME('%w', date) as integer)
when 0 then 'Sunday'
when 1 then 'Monday'
when 2 then 'Tuesday'
when 3 then 'Wednesday'
when 4 then 'Thursday'
when 5 then 'Friday'
else 'Saturday' end as weekday

FROM marketing
GROUP BY weekday
ORDER BY AVG(clicks) DESC
''')

AVG(clicks)   weekday
0  1423.933333  Saturday

The best day of the weekn to run ads is Saturday which has the highest average of 1423 clicks 